### PR TITLE
Replicated Sharing Tensor (RST): secret sharing, reconstruction, and linear operations

### PR DESCRIPTION
**Pull request by abogaziah**
## Description
This PR will implement the basic operations of the replicative secret sharing scheme in order to be able to implement falcon protocol
for more details: #3852 

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

Milestones: 
- [x] Secret sharing
- [x] Reconstruction
- [x] Public and private addition / sub

### CODE REVIEW

File: syft/frameworks/torch/tensors/interpreters/replicative_shared.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            shares.append(torch.tensor(random.randrange(self.ring_size)))
+        shares.append(torch.tensor(secret - sum(shares) % self.ring_size))
+        return shares
+
+    @staticmethod
+    def distribute_shares(workers, shares):
+        shares_locations = {}
+        assert len(workers) == len(shares)
+        for i in range(len(shares)):
+            pointer1 = shares[i].send(workers[i])
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicative_shared.py:33]: Q: Shouldn't here the party[i] hold shares i-1 and i + 1?

Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicative_shared.py:33]: No, it should know shares i and i+1. if it knows i-1 and i+1 it can reconstruct the secret.
section 3.1 of the paper states: "(x1, x2)
is held by P1, (x2, x3) by P2, and (x3, x1) by P3"

File: syft/frameworks/torch/tensors/interpreters/replicative_shared.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        self.ring_size = 2 ** 5
+
+    def share_secret(self, secret, workers):
+        number_of_shares = len(workers)
+        shares = self.generate_shares(secret, number_of_shares)
+        shares_locations = self.distribute_shares(workers, shares)
+        self.child = shares_locations
+        return self
+
+    def generate_shares(self, secret, number_of_shares=3):
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicative_shared.py:21]: MAYBE: make this ```staticmethod``` and test it with more shares?

Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicative_shared.py:21]: I don't think this can be a static method because it uses an instance variable (self.ring_size), but we can still test it on more shares

Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/replicative_shared.py:21]: PICKY: This is meant to be called only from inside the ```ReplicatedSharingTensor``` class, right? (in that case maybe we can prefix the method name with ```_``` such that we know it is internal?

**Comment by gmuraru** [comment]
@abogaziah I removed some points from this task and created other issues for each one individually.
They can be seen in the [FALCON Milestone](https://github.com/OpenMined/PySyft/milestone/13)

File: syft/frameworks/torch/tensors/interpreters/replicative_shared.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        pointers = list(shares_map[players[0]])
+        pointers.append(shares_map[players[1]][1])
+        return pointers
+
+    def __sum_shares(self, shares):
+        return sum(shares) % self.ring_size
+
+    def __map_modular_to_real(self, mod_number):
+        """In a modular ring, a number x is mapped to a negative
+         real number ]0,-∞[ iff x > ring_size/2 """
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicative_shared.py:74]: TYPO: ```[``` should be ```]``` and ```]``` should be ```[```

File: syft/frameworks/torch/tensors/interpreters/replicative_shared.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import syft
+from operator import add, sub
+
+
+class ReplicatedSharingTensor(AbstractTensor):
+    def __init__(
+        self, shares_map=None, owner=None, id=None, tags=None, description=None,
+    ):
+        super().__init__(id=id, owner=owner, tags=tags, description=description)
+        self.child = shares_map
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicative_shared.py:13]: MAYBE: The ```shares_map``` should be only ```shares```. We can get the ```player``` for a share (or the owner) using the ```location``` attribute for the share

Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/replicative_shared.py:13]: This is how it is done in the ```AdditiveSharingTensor```

File: syft/frameworks/torch/tensors/interpreters/replicative_shared.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        if not self.verify_matching_players(secret):
+            raise ValueError("Shares must be distributed among same parties")
+        z = {}
+        x, y = self.get_shares_map(), secret.get_shares_map()
+        for player in x.keys():
+            z[player] = (operator(x[player][0], y[player][0]), operator(x[player][1], y[player][1]))
+        return ReplicatedSharingTensor(z)
+
+    def public_linear_operation(self, plain_text, operator):
+        players = self.get_players()
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/replicative_shared.py:119]: MAYBE: If you use ```shares``` and not ```shares_map``` this should simply be ```[player for player in self.shares.location]```

File: syft/frameworks/torch/tensors/interpreters/replicative_shared.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        number_of_shares = len(workers)
+        workers = self.__arrange_workers(workers)
+        shares = self.generate_shares(secret, number_of_shares)
+        shares_map = self.__distribute_shares(workers, shares)
+        self.child = shares_map
+        return self
+
+    @staticmethod
+    def __arrange_workers(workers):
+        """ having local worker in index 0 saves one communication round"""
```
[the remark this answers is not in the document]
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicative_shared.py:26]: because the shares are local, you won't need to communicate (which is an expensive operation) with another worker to get them thus saved a communication round

File: syft/frameworks/torch/tensors/interpreters/native.py [hunk 7]
```diff
@@ -909,6 +909,11 @@ def share(
             requires_grad (bool): Should we add AutogradTensor to allow gradient computation,
                 default is False.
         """
+        if protocol == "falcon":
+            shared_tensor = syft.ReplicatedSharingTensor(owner=self.owner).share_secret(
```
Comment (gmuraru) [inline syft/frameworks/torch/tensors/interpreters/native.py:913]: Q: I am thinking if we can keep a ```dict``` for those.
Like:
```dict = {
  "falcon": syft.ReplicatedSharingTensor,
  "snn"    : syft.AdditiveSharingTensor
}
```

File: syft/frameworks/torch/tensors/interpreters/replicative_shared.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
+import torch
```
[the remark this answers is not in the document]
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicative_shared.py:1]: according to clean coding principles, comments are very bad practice, they often are misleading, don't change with code. I made the code self-documented so comments are as least as possible, I tried adding comments but it was much harder to read with comments for me than without it

File: syft/frameworks/torch/tensors/interpreters/replicative_shared.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+from syft.generic.abstract.tensor import AbstractTensor
+import random
+import syft
+from operator import add, sub
+
+
+class ReplicatedSharingTensor(AbstractTensor):
+    def __init__(
+        self, shares_map=None, owner=None, id=None, tags=None, description=None,
```
[the remark this answers is not in the document]
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicative_shared.py:10]: I haven't faced a scenario where I need to specify the dtype in this class so far, we can add it if we need it anytime

File: syft/frameworks/torch/tensors/interpreters/replicative_shared.py (hunk 3 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicative_shared.py:74]: that's not actually a typo, open bracket means the set doesn't include that number. negative numbers are everything between 0 and - infinity not including them, I can remove it if it's confusing, "negative real number" is kinda sufficient

File: syft/frameworks/torch/tensors/interpreters/replicative_shared.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    ):
+        super().__init__(id=id, owner=owner, tags=tags, description=description)
+        self.child = shares_map
+        self.ring_size = 2 ** 5
+
+    def share_secret(self, secret, workers):
+        number_of_shares = len(workers)
+        workers = self.__arrange_workers(workers)
```
[the remark this answers is not in the document]
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicative_shared.py:18]: let's add tests and see :"D

File: syft/frameworks/torch/tensors/interpreters/native.py (hunk 7 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/native.py:913]: what would be the use-case of that dict? I don't get it

File: syft/frameworks/torch/tensors/interpreters/replicative_shared.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+    def generate_shares(self, plain_text, number_of_shares=3):
+        shares = []
+        for _ in range(number_of_shares - 1):
+            shares.append(torch.tensor(random.randrange(self.ring_size)))
+        shares.append(torch.tensor((plain_text - sum(shares)) % self.ring_size))
+        return shares
+
+    @staticmethod
+    def __distribute_shares(workers, shares):
```
[the remark this answers is not in the document]
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicative_shared.py:40]: " A single leading underscore means you're not supposed to access it "from the outside" -- two leading underscores (w/o trailing underscores) carry the message even more forcefully"
https://www.geeksforgeeks.org/private-methods-in-python/
https://stackoverflow.com/questions/1547145/defining-private-module-functions-in-python
http://www.faqs.org/docs/diveintopython/fileinfo_private.html

File: syft/frameworks/torch/tensors/interpreters/replicative_shared.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicative_shared.py]: done! I added tests and it works correctly now

File: syft/frameworks/torch/tensors/interpreters/native.py (hunk 7 above)
Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/native.py:913]: Dict are better to dynamically add options if needed, and are faster to access instead of if / else-ing on all optiosn

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+    @staticmethod
+    def __arrange_workers(workers):
+        """ having local worker in index 0 saves one communication round"""
+        if syft.hook.local_worker in workers:
+            workers.remove(syft.hook.local_worker)
+            workers = [syft.hook.local_worker] + workers
+        return workers
+
+    def generate_shares(self, plain_text, number_of_shares=3):
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/replicated_shared.py:32]: I recommend taking example on https://github.com/OpenMined/PySyft/blob/4518a44700cefcd6efd4b43dbf7bf8ae79c52afb/syft/frameworks/torch/tensors/interpreters/additive_shared.py#L334
i.e. remove the noise you've added to the next share instead of removing the sum of all on the last share

Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:32]: What is the difference? in other words, why would you choose this over that?

**Comment by gmuraru** [comment]
I think it looks great as a first PR that enables us to use RSS.
What I think we should implement more is - have the possibility to do operations using "+", "-", etc (I do not know if that is possible for the moment) - There might be needed some ```@overloaded``` work

**Comment by abogaziah** [comment]
@gmuraru done :)

File: syft/frameworks/torch/tensors/interpreters/native.py (hunk 7 above)
Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/native.py:913]: @abogaziah are you making the changes in this PR or we can create a subsequent ticket to do the changes in another issue (we can also mark it as good first issue :D)

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 12 above)
Comment (gmuraru) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:32]: ping @LaRiffle

File: syft/frameworks/torch/tensors/interpreters/native.py (hunk 7 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/native.py:913]: I think it needs to be done later, because sharing doesn't work the same way in both cases for now, (AST is more complex) maybe we should do that after refactoring AST

File: syft/frameworks/torch/tensors/interpreters/replicated_shared.py (hunk 12 above)
Comment (abogaziah) [reply syft/frameworks/torch/tensors/interpreters/replicated_shared.py:32]: I did, I'm looking at how crypten handles it

File: syft/frameworks/torch/tensors/interpreters/native.py (hunk 7 above)
Comment (LaRiffle) [reply syft/frameworks/torch/tensors/interpreters/native.py:913]: Sounds good! There is also some refacto going on in the AST file in #3924 :)
